### PR TITLE
Add CAPTURE_FOR_ERROR to print info on failure

### PR DESCRIPTION
**Pull request by wthrowe**
Example:
```c++
    const Scalar<DataVector> scalar{DataVector{0.0, 1.0, 5.0, 4.0}};
    const int integer = 12;
    CAPTURE_FOR_ERROR(scalar);
    CAPTURE_FOR_ERROR(integer);
    ERROR("Boom");
```
Leads to:
```
[backtrace and stuff]

Boom

Captured variables:
scalar = T()=(0,1,5,4)
integer = 12
############ ERROR ############
```

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Utilities/ErrorHandling/CaptureForError.hpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+  // Capturing an rvalue would lead to a dangling reference.
+  CaptureForError(const char* /*name*/, const T&& /*capture*/) = delete;
+
+  ~CaptureForError() noexcept(false) {
+    CaptureList::the_list().deregister_self(this);
+  }
+
+  void print(std::ostream& stream) const override {
+    stream << name_ << " = " << capture_ << "\n";
```
Comment (nilsdeppe) [inline src/Utilities/ErrorHandling/CaptureForError.hpp:71]: maybe add `using ::operator<<;` to deal with STL as best as we can?

**Review by nilsdeppe — APPROVED** [review]
This is awesome :D Thank you for implementing this so quickly! One small question, squash immediately if you agree :)

**Comment by nilsdeppe** [comment]
Marked priority because this will help tremendously with debugging BNS simulations

File: src/Utilities/ErrorHandling/CaptureForError.hpp (hunk 1 above)
Comment (nilsdeppe) [reply src/Utilities/ErrorHandling/CaptureForError.hpp:71]: And, maybe more important, call `disable_floating_point_exceptions()` (and re-enable)? Or maybe that's handle by `ERROR` itself

Comment (wthrowe) [reply src/Utilities/ErrorHandling/CaptureForError.hpp:71]: `using ::operator<<;` won't compile unless I include something that defines such a function.  I can add it with an include of `StdHelpers.hpp` if you think that's a good idea.

`ERROR` and `ASSERT` have FPEs disabled when they call this.

Comment (nilsdeppe) [reply src/Utilities/ErrorHandling/CaptureForError.hpp:71]: Ah okay, nah let's just leave as-is... Thanks!

**Pull request merged**